File: frontend/src/types/workflow.ts
/**
 * TypeScript type definitions for workflow schemas.
 * Mirrors backend/app/schemas/workflow.py
 */

export interface Position {
  x: number;
  y: number;
}

export interface WorkflowNodeData {
  cube_id: string;
  params: Record<string, unknown>;
}

export interface WorkflowNode {
  id: string;
  type: string;
  position: Position;
  data: WorkflowNodeData;
}

export interface WorkflowEdge {
  id: string;
  source: string;
  target: string;
  sourceHandle: string | null;
  targetHandle: string | null;
}

export interface WorkflowGraph {
  nodes: WorkflowNode[];
  edges: WorkflowEdge[];
}

export interface WorkflowResponse {
  id: string;
  name: string;
  graph_json: WorkflowGraph;
  created_at: string;
  updated_at: string;
}

File: backend/app/schemas/workflow.py
import uuid
from datetime import datetime

from pydantic import BaseModel, Field


class Position(BaseModel):
    x: float
    y: float


class WorkflowNodeData(BaseModel):
    cube_id: str
    params: dict = Field(default_factory=dict)


class WorkflowNode(BaseModel):
    id: str
    type: str = "cube"
    position: Position
    data: WorkflowNodeData


class WorkflowEdge(BaseModel):
    id: str
    source: str
    target: str
    sourceHandle: str | None = None
    targetHandle: str | None = None


class WorkflowGraph(BaseModel):
    nodes: list[WorkflowNode] = Field(default_factory=list)
    edges: list[WorkflowEdge] = Field(default_factory=list)


class WorkflowCreate(BaseModel):
    name: str = "Untitled Workflow"
    graph_json: WorkflowGraph = Field(default_factory=WorkflowGraph)


class WorkflowUpdate(BaseModel):
    name: str | None = None
    graph_json: WorkflowGraph | None = None


class WorkflowResponse(BaseModel):
    id: uuid.UUID
    name: str
    graph_json: WorkflowGraph
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}
